src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -248,12 +248,13 @@ impl Validator<'mir, 'tcx> {
             return;
         }
 
-        // If an operation is supported in miri (and is not already controlled by a feature gate) it
-        // can be turned on with `-Zunleash-the-miri-inside-of-you`.
-        let is_unleashable = O::IS_SUPPORTED_IN_MIRI && O::feature_gate().is_none();
+        // If an operation is supported in miri it can be turned on with
+        // `-Zunleash-the-miri-inside-of-you`.
+        let is_unleashable = O::IS_SUPPORTED_IN_MIRI;
 
         if is_unleashable && self.tcx.sess.opts.debugging_opts.unleash_the_miri_inside_of_you {
-            self.tcx.sess.span_warn(span, "skipping const checks");
+            self.tcx.sess.delay_span_bug(self.span, "skipping const checks is for testing error paths only");
+            self.tcx.sess.span_warn(self.tcx.def_span(self.def_id), "skipping const checks");
```

Comment:
I feel like we don't need both `delay_span_bug` and `warn`; would you be okay removing the warning then?

Reply:
I'm not sure we need the `delay_span_bug` at all here. In case the test doesn't need const checks, this won't actually trigger the delay span bug, which is why `read_from_static` doesn't trigger it.

I'll let you know of my findings (working on a PR about the delay span bug thing right now)

Reply:
We have more run-pass tests. E.g. https://github.com/rust-lang/rust/blob/8417d68de5e063426ab6bb7f383df6117d1beeed/src/test/ui/rfc-2091-track-caller/caller-location-fnptr-rt-ctfe-equiv.rs needs unleash in order to check something at runtime.

So we can't use delay span bug to achieve "checking that the only skips are skipping feature gates". This does in fact need a separate system.

Reply:
> I'm not sure we need the delay_span_bug at all here.

This is the only place where the flag makes any behavioral difference. So if this `delay_span_bug` doesn't fire, then sure the flag is unnecessary, but it also does not disable any feature checks or so, it just does nothing.

Reply:
> We have more run-pass tests

That looks more like we should just add a feature gate for `const fn` pointers, and then use that?

Reply:
> That looks more like we should just add a feature gate for const fn pointers, and then use that?

That requires nontrivial parser changes, but yes, we should.

Reply:
> This is the only place where the flag makes any behavioral difference. So if this delay_span_bug doesn't fire, then sure the flag is unnecessary, but it also does not disable any feature checks or so, it just does nothing.

Well, there are multiple `"skipping const check"` in the source, and the only thing it checks now is whether unleash tests that actually unleash something cause an error, which I don't believe is what @ecstatic-morse had in mind.

Reply:
> That requires nontrivial parser changes, but yes, we should.

Well, that code clearly already works without any parser changes. So adding a feature gate just requires implementing `feaure_gate` for `FnCallIndirect`, doesn't it?

Reply:
> Well, there are multiple "skipping const check" in the source, and the only thing it checks now is whether unleash tests that actually unleash something cause an error, which I don't believe is what @ecstatic-morse had in mind.

They [said themselves](https://github.com/rust-lang/rust/pull/71631#issuecomment-622013979) that it "would work".

Your proposal of somehow detecting whether the test could work without feature gates, doesn't really match the main use-case for this flag which is to *disable all static checks* (no matter whether they are stable or unstable or whatever) so that we can test the dynamic ones. I just don't want to have to care what the static checks are saying for this code. The flag means "static checks, shut up". Your proposal means I still have to care about the static checks.

Reply:
> Well, that code clearly already works without any parser changes. So adding a feature gate just requires implementing feaure_gate for FnCallIndirect, doesn't it?

There's no way we'll ever stabilize it in a way that will work that way. Feature gates are something that is only unstable because it's baking on nightly and maybe has a few bugs or because it's a compiler internal thing needed for implementing libstd. But at least it has a tracking issue. Unleash was to me always for things that didn't have a way forward to stabilization at the time when I added the unleash path.

Otherwise we could add feature gates for everything, even `UnsafeCell` behind references in constants. I'm ok with changing the "rules" (which are vague, I admit), but we need a line somewhere that differentiates "this is unsound" from "this is unstable".

Reply:
> They said themselves that it "would work".

From the context (copied below), I think there was some confusion about what effect that `delay_span_bug` would have.

>> we can make sure that when a test specifies unleash, it needs to actually unleash anything in order to pass.
>
> This is the thing I want to enforce. The goal is to prevent you or (more importantly) others from using `-Zunleash` as a replacement for `#![feature(many,different,things)]` either in-tree or downstream.

Reply:
More concretely, I don't want people using `-Zunleash` to get around adding a [long list of feature gates](https://github.com/ecstatic-morse/const-phf/blob/1a8768ea97d297b47873b8e753779e521df21356/src/lib.rs#L1-L12) to make their test or ecosystem code work. I actually wasn't thinking about the case where tests accidentally add `-Zunleash` even when it does literally nothing. That's not really tied to my original concern, since they used all the correct feature gates. It does seem like we should warn in this case, though.

Always making `-Zunleash` cause an error prevents people from using it to avoid feature gates, since their code just won't work. FWIW, I think the status quo achieves this in a more straightforward way, but that seems to be a no-go for some reason. However, `delay_span_bug` means that we can't have `run-pass` unleash tests. I think Ralf thought there was no use case for them, but that seems to be incorrect?

Reply:
> More concretely, I don't want people using -Zunleash to get around adding a long list of feature gates to make their test or ecosystem code work.

I consider that a non-issue outside of rustc's test suite, as specifying `-Z` command line flags is not something to easily do outside of manually invoking nighlty `rustc`. Even if you use `RUSTFLAGS`, that will just be a mess, give you loads of `skipping const checks` warnings and the flag's documentation states that it is unsound!

> Always making -Zunleash cause an error prevents people from using it to avoid feature gates, since their code just won't work.

I don't think people are using it in practice (https://github.com/search?p=7&q=%22unleash+the+miri+inside+of+you%22&type=Code) even if https://github.com/rust-embedded/blog/blob/bfc76e942f890142c2be9edbe0d15f0c829eff39/content/2019-02-20-newsletter-15.md#L73 and https://github.com/rust-embedded/wg/blob/056080eaf325d6f81413a172f2bd635fafa39ec4/minutes/2019-02-05.irc.log#L142 seem to suggest that potentially people may use it to get around limitations without realizing it is unsound, but that's been a year ago without any usage of said flag on github.

> FWIW, I think the status quo achieves this in a more straightforward way, but that seems to be a no-go for some reason.

I believe (correct me @RalfJung if I'm misinterpreting), that the issue is that while it's certainly possible to add feature flags, it's a pointless boilerplate excercise, because for the test the stability of any features is irrelevant, it is supposed to be testing the miri-engine. Technically we could just wrap the entire const check infrastructure in a single `if unleash` check instead of doing it per gate (but that would stop giving us warnings and stuff).

Maybe a middle ground would be to replace the "skipping const checks" warning with a "skipping feature gate check for `some_feature_name`" in case it's a feature gate check? Then it would still be obvious what's going on and CI would still break for that test if something about the feature gate changes, but that breakage could be blessed away.

Reply:
> without realizing it is unsound

It hopefully isn't unsound. ;)

> that the issue is that while it's certainly possible to add feature flags, it's a pointless boilerplate excercise, because for the test the stability of any features is irrelevant, it is supposed to be testing the miri-engine

That summarizes my point, yes.

I have an idea for yet another middle ground proposal that hopefully keeps all existing tests (except for the silly one) working and makes everyone happy... let's see if that works.

Reply:
> Maybe a middle ground would be to replace the "skipping const checks" warning with a "skipping feature gate check for some_feature_name" in case it's a feature gate check? Then it would still be obvious what's going on and CI would still break for that test if something about the feature gate changes, but that breakage could be blessed away.

That would be fine for me, too. It might be preferable as it does not require the hacky `impl Drop for Session`.